rs/p2p/artifact_manager/src/lib.rs

```diff
@@ -224,18 +224,25 @@ fn process_messages<Artifact: IdentifiableArtifact + 'static>(
             }
         };
         let ChangeResult {
-            artifacts_with_opt,
-            purged,
+            mutations,
             poll_immediately,
         } = metrics
             .with_metrics(|| client.process_changes(time_source.as_ref(), batched_artifact_events));
-        for artifact_with_opt in artifacts_with_opt {
-            let _ = send_advert.blocking_send(ArtifactProcessorEvent::Artifact(artifact_with_opt));
-        }
 
-        for advert in purged {
-            let _ = send_advert.blocking_send(ArtifactProcessorEvent::Purge(advert));
+        // We must first send the addition to the replication manager because in theory in one batch we can have both an addition and removal of the same artifact.
+        let mut purged = vec![];
+        for mutation in mutations {
+            match mutation {
+                ArtifactMutation::Insert(artifact_with_opt) => {
+                    let _ = send_advert.blocking_send(ArtifactMutation::Insert(artifact_with_opt));
+                }
+                ArtifactMutation::Remove(id) => purged.push(id),
+            };
         }
+        for id in purged {
+            let _ = send_advert.blocking_send(ArtifactMutation::Remove(id));
+        }
+
```

Comment:
You could argue that with this API (`Vec<Mutation>`) it is the job of consensus to order the artifacts in a way that makes sense. The reordering here is a little surprising now whereas before it made sense to send inserts first.

Reply:
I agree it does make more sense. I made the change.

@eichhorl can you PTAL at the MR and confirm it is fine with you as well.